### PR TITLE
ref code: do not clear the secnonce in deterministic_sign 

### PR DESCRIPTION
**Pull request by jonasnick**

### CODE REVIEW

**Comment by jonasnick** [comment]
Addresses https://github.com/jonasnick/bips/pull/11#discussion_r955136137

File: bip-musig2/reference.py [hunk 1]
```diff
@@ -1,3 +1,9 @@
+# BIP-musig2 reference implementation
```
Comment (real-or-random) [inline bip-musig2/reference.py:1]: nit: I bet we'll forget to update this when we get a number :)

**Review by real-or-random — APPROVED** [review]
ACK d59feeb26b7d48aa40829dd14f465a0f2e03cdb3 nice :)

**Review by robot-dreams — COMMENTED** [review]
Concept ACK

To clarify, what is the primary goal?

1. Ensure that `secnonce` cannot be reused by a careless user (e.g. it's good enough if a `secnonce` variable falls out of scope and gets garbage collected later)
2. Ensure that `secnonce` doesn't linger in memory, even if it isn't reachable from any code (e.g. overwrite the bytes immediately)

If our goal is (1) then I think this PR is good to go. If it's (2) we should clarify further, and maybe point out `bytes` isn't a good choice (it's immutable)

**Comment by real-or-random** [comment]
Hm now I see that my comment in #11 was slightly mislead. I thought this was just about clearing secrets from the stack but this is different here.  The text says this:

```
To avoid accidental reuse of secnonce, an implementation may securely erase the secnonce argument by overwriting it with 64 zero bytes after it has been read by Sign'
```

So with this in mind, I think both (1) and (2) are a good goals. 

Would it then be a good idea to change `secnonce` to a `bytearray`?  https://docs.python.org/3.8/library/stdtypes.html#bytearray
I believe this would be a solution that is also appropriate for python, and it would catch both (1) and (2) in normal Sign. (The situation for DeterministicSign is a little different because there `secnone` is anyway only computed inside the scope of the algorithm.)

I tend to say that this is not overkill even in a reference implementation that isn't constant-time etc (because we stress this overwriting aspect in the BIP text explicitly) .

Could we change the text to say Sign itself should overwrite the argument? Then we could even change the pseudocode, why not.

(By the way, the text should be updated because it mentions `Sign'`, just a typo... )

**Comment by jonasnick** [comment]
I hadn't considered using a bytearray and overwriting it in sign like we do in libsecp-zkp and suggest in the text. That's much better. Force pushed a commit that does that.

**Comment by jonasnick** [comment]
Added change log entry

File: bip-musig2/gen_vectors_helper.py [hunk 2]
```diff
@@ -127,7 +127,8 @@ def sig_agg_vectors():
         session_ctx = SessionContext(aggnonce, vec_pubkeys, vec_tweaks, is_xonly, msg)
 
         for j in range(len(key_indices)):
-            psigs[psig_indices[j]] = sign(secnonces[nonce_indices[j]], sk[key_indices[j]], session_ctx)
+            secnonce_tmp = bytearray(secnonces[nonce_indices[j]])
+            psigs[psig_indices[j]] = sign(secnonce_tmp, sk[key_indices[j]], session_ctx)
```
Comment (real-or-random) [inline bip-musig2/gen_vectors_helper.py:134]: nit: add the "don't copy" warning also here?

File: bip-musig2/reference.py [hunk 3]
```diff
@@ -830,9 +830,9 @@ def test_sign_and_verify_random(iters: int) -> None:
 
         session_ctx = SessionContext(aggnonce, pubkeys, tweaks, is_xonly, msg)
         psig_1 = sign(secnonce_1, sk_1, session_ctx)
-        # Clear the secnonce after use to avoid accidental reuse
-        secnonce_1 = bytes(64)
         assert partial_sig_verify(psig_1, pubnonces, pubkeys, tweaks, is_xonly, msg, 0)
+        # An exception is thrown if secnonce_1 is accidentally reused
+        assert_raises(ValueError, lambda: sign(secnonce_1, sk_1, session_ctx), lambda e: True)
```
Comment (real-or-random) [inline bip-musig2/reference.py:836]: That's good. Should we have signing test vectors for when k1 = 0 or k2 = 0?

**Review by real-or-random — COMMENTED** [review]
Nice, that's much cleaner.

File: bip-musig2/reference.py (hunk 3 above)
Comment (jonasnick) [reply bip-musig2/reference.py:836]: Good idea, done.

File: bip-musig2/gen_vectors_helper.py (hunk 2 above)
Comment (jonasnick) [reply bip-musig2/gen_vectors_helper.py:134]: done

**Review by real-or-random — APPROVED** [review]
utACK 0682f0920575cc71cee52620d7d93be5d18c7a96

**Comment by robot-dreams** [comment]
ACK 0682f0920575cc71cee52620d7d93be5d18c7a96, indeed much clearer!

**Pull request merged**